### ISSUE TITLE
"This is an internal Run bug" report

### ISSUE THREAD

**Issue by david0178418** [runonbitcoin/sdk#5]
I received this error message when attempting to sync a jig.

The in-memory state of the jig instance was at this location: d54b774e8983ecb1b2c1d72cc2877c391ac007b56984595c8be9beac753e73dd_o1

Another client updated that instance here: 6f2e5c77a01840599406dc51d8e2696b1c374ae9e1bef5e76b82909fbc029532_o1

After the update, I ran "sync" to trip the below message

EDIT: Forgot to paste in that "time" string.  Not sure if it's necessary, but I copied it out of the image to save some time in case it is. (also, just be aware of that fact in case there's a typo)  time://557c2c06c05f8efb87d5d5d0ba06c85a0ae81a543fc3be5bb846995c7efbf03f

![image](https://user-images.githubusercontent.com/128739/109357038-2d1fc500-7847-11eb-9721-f934a472893b.png)

**Comment by brentongunning** [comment]
Thanks David for filing this! Yes, this is indeed looks like a Run bug.

In the mean time, you can use this LocalCache implementation with that check removed. It should be fine for now.

```
class LocalCache {
  constructor (options = {}) {
    this._map = new Map()
    this._sizeBytes = 0
    this._maxSizeBytes = this._parseMaxSizeMB(options.maxSizeMB) * 1000 * 1000
  }

  get maxSizeMB () {
    return this._maxSizeBytes / 1000 / 1000
  }

  set maxSizeMB (value) {
    this._maxSizeBytes = this._parseMaxSizeMB(value) * 1000 * 1000
    this._reduceToFit()
  }

  get sizeBytes () { return this._sizeBytes }

  async get (key) {
    const had = this._map.has(key)
    const value = this._map.get(key)

    if (had) {
      // Bump the entry to the top
      this._map.delete(key)
      this._map.set(key, value)

      return value
    }
  }

  async set (key, value) {
    const had = this._map.has(key)
    const previous = this._map.get(key)

    // Bump the entry to the top, or set the new value
    this._map.delete(key)
    this._map.set(key, value)

    if (had) return

    this._sizeBytes += LocalCache._estimateSize(key)
    this._sizeBytes += LocalCache._estimateSize(value)

    this._reduceToFit()
  }

  clear () {
    this._map.clear()
    this._sizeBytes = 0
  }

  _reduceToFit () {
    while (this._sizeBytes > this._maxSizeBytes) {
      const oldestKey = this._map.keys().next().value
      const oldestValue = this._map.get(oldestKey)
      this._map.delete(oldestKey)
      this._sizeBytes -= LocalCache._estimateSize(oldestKey)
      this._sizeBytes -= LocalCache._estimateSize(oldestValue)
    }
  }

  static _estimateSize (value) {
    // Assumes only JSON-serializable values
    // Assume each property has a 1 byte type field, and pointers are 4 bytes
    // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures
    switch (typeof value) {
      case 'undefined': throw new Error('Cannot cache undefined')
      case 'boolean': return 5
      case 'number':
        if (Number.isFinite(value)) return 9
        throw new Error(`Cannot cache number: ${value}`)
      case 'string': return value.length * 2 + 1
      case 'object': {
        if (!value) return 5
        const keys = Object.keys(value)
        let size = 1 + keys.length * 4
        keys.forEach(key => {
          size += LocalCache._estimateSize(key)
          size += LocalCache._estimateSize(value[key])
        })
        return size
      }
      case 'function': throw new Error(`Cannot cache function: ${value}`)
      case 'symbol': throw new Error(`Cannot cache symbol: ${value}`)
    }
  }

  function _parseMaxSizeMB (maxSizeMB) {
    if (typeof maxSizeMB === 'undefined') return 10
    if (typeof maxSizeMB === 'number' && !Number.isNaN(maxSizeMB) && maxSizeMB >= 0) return maxSizeMB
    throw new Error(`Invalid maxSizeMB: ${maxSizeMB}`)
  }
}
```

Plug that into your Run instance via: `new Run({ cache: new LocalCache() })`

**Comment by brentongunning** [comment]
Thanks @david0178418. This should be fixed in 0.6.18. Please re-open if you still see it.

https://github.com/runonbitcoin/sdk/commit/fb26e69607ede25c69995b4e63c2a8e25767ac35
